fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java

```diff
@@ -879,26 +1119,157 @@ private ScalarOperator getMVCompensationPredicate(RewriteContext rewriteContext,
             }
         }
 
-        return MvUtils.canonizePredicate(Utils.compoundAnd(equalPredicates, otherPredicates));
+        ScalarOperator compensationPredicate = MvUtils.canonizePredicate(Utils.compoundAnd(equalPredicates, otherPredicates));
+        if (compensationPredicate == null) {
+            return null;
+        }
+        return addJoinDerivePredicate(rewriteContext, rewriter, mvColumnRefToScalarOp, compensationPredicate);
+    }
+
+    private ScalarOperator addJoinDerivePredicate(
+            RewriteContext rewriteContext,
+            ColumnRewriter rewriter,
+            Map<ColumnRefOperator, ScalarOperator> mvColumnRefToScalarOp,
+            ScalarOperator compensationPredicate) {
+        // unnecessary to add the derived compensation predicates if compensationPredicate has related null rejecting predicate
+        List<ScalarOperator> predicates = Utils.extractConjuncts(compensationPredicate);
+        List<ScalarOperator> derivedPredicates = Lists.newArrayList();
+        for (JoinDeriveContext joinDeriveContext : mvRewriteContext.getJoinDeriveContexts()) {
+            if (joinDeriveContext.getMvJoinType().isInnerJoin() && joinDeriveContext.getQueryJoinType().isSemiJoin()) {
+                continue;
+            }
+            Optional<ScalarOperator> derivedPredicateOpt = null; 
+            if (joinDeriveContext.getMvJoinType().isLeftOuterJoin() && joinDeriveContext.getQueryJoinType().isInnerJoin()) {
+                List<ColumnRefOperator> rightJoinColumns = joinDeriveContext.getRightJoinColumns();
+                derivedPredicateOpt = getDerivedPredicate(rewriteContext,
+                        rewriter, mvColumnRefToScalarOp, rightJoinColumns, predicates, true, true);
+            } else if (joinDeriveContext.getMvJoinType().isLeftOuterJoin()
```

Comment:
looks like it's better to use a Map<Pair<MVJoinType, QueryJoinType>, Lambda> ?

Reply:
why?